__tests__/Datepicker.js

```diff
@@ -1,26 +1,74 @@
 import React from 'react'
 import Datepicker from '../src/Datepicker'
 import Calendar from '../src/Calendar'
-import { mount } from 'enzyme';
-import { format, getDaysInMonth } from 'date-fns'
-
-let datepicker
-
-beforeEach(() => {
-  datepicker = mount(<Datepicker />)
-});
+import { mount, shallow } from 'enzyme';
+import addDays from 'date-fns/add_days'
 
 test('calendar should be hidden by default', () => {
+  const datepicker = mount(<Datepicker />)
   expect(datepicker.state().showCalendar).toBeFalsy()
 });
 
+test('input should be readOnly by default', () => {
+  const datepicker = shallow(<Datepicker />)
+  const input = datepicker.find('input')
+
+  expect(input.props().readOnly).toBeTruthy()
+  
+  input.simulate('change', {target: {value: '10/10/2010'}});
+  expect(input.props().value).not.toBe('10/10/2010')
+})
+
+test('if passed enableTextInput, input should accept typed text', () => {
+  const datepicker = mount(<Datepicker dateFormat="MM/DD/YYYY" enableTextInput />)
+  const input = datepicker.find('input')
+  
+  expect(input.props().readOnly).toBeFalsy()
+  
+  input.simulate('change', {target: {value: '10/10/2010'}});
+  
+  expect(input.props().value).toBe('10/10/2010')
+})
+
+test('if passed enableTextInput, state should only change on valid date', () => {
+  const startDate = new Date('2020-01-04T06:00:00.000Z')
+  const datepicker = mount(<Datepicker dateFormat="MM/DD/YYYY" startDate={startDate} enableTextInput/>)
+  const input = datepicker.find('input')
+  
+  input.simulate('change', {target: {value: '10/10/201a'}});
+  expect(datepicker.state().selectedDate).toBe(startDate)
+  
+  input.simulate('change', {target: {value: '10/10/2020'}});
+  expect(datepicker.state().selectedDate).toBe('10/10/2020')
+
+  input.simulate('change', {target: {value: 'ZZ/10/2020'}});
+  expect(datepicker.state().selectedDate).toBe('10/10/2020')
+})
+
+test('if passed enableTextInput, should respect mindate & maxdate', () => {
+  const startDate = new Date('2020-01-04T06:00:00.000Z')
+  const datepicker = mount(<Datepicker dateFormat="MM/DD/YYYY" startDate={startDate} minDate={startDate} maxDate={addDays(startDate, 10)} enableTextInput/>)
+  const input = datepicker.find('input')
+  
+  input.simulate('change', {target: {value: '01/10/2020'}});
+  expect(datepicker.state().selectedDate).toBe('01/10/2020')
+  
+  input.simulate('change', {target: {value: '01/20/2020'}});
```

Comment:
changing to `01/19/2020` would work though right?

Reply:
No, the startDate is actually 1/4/2020, so 1/14 and before would work.

Reply:
It's generally best practice to test the boundary conditions, so I'd suggest using 1/15 here.